### PR TITLE
[ASTGen] Infrastructure for implementing ASTGen incrementally

### PR DESCRIPTION
**Pull request by rintaro**
* `ASTGenVisitor` has a reference to a legacy C++ Parser configured for ASTGen.
* When `ASTGenVisitor` encounters a AST node that hasn't been migrated, call the C++ `Parser::parse(Decl|Stmt|Expr|Type)` to parse the position using the legacy parser.
* If the legacy parser is configured for ASTGen, it calls ASTGen's  `swift_ASTGen_build(Decl|Stmt|Expr|Type)` for each ASTNode parsing (unless the call is not directly from the ASTGen.)

rdar://117151886

### CODE REVIEW

File: include/swift/Parse/Parser.h
Comment (ahoppen) [inline include/swift/Parse/Parser.h:163]: Why did you choose to put this behind a `#if`? Seems like a potential source of issues in the future if you use it from somewhere that isn’t guarded by `SWIFT_BUILD_SWIFT_SYNTAX` and that you won’t see locally if you have `SWIFT_BUILD_SWIFT_SYNTAX` enabled.

File: include/swift/Parse/Parser.h
Comment (ahoppen) [inline include/swift/Parse/Parser.h:966]: Could you add a comment what `fromASTGen` does? Similar for the other functions that have `fromASTGen`.

---

Also, all the other parameter start with uppercase.

File: include/swift/Parse/Parser.h
Comment (ahoppen) [inline include/swift/Parse/Parser.h:2046]: Just reading the function declaration, I’m thinking: Which syntax tree is this parsing from? I think a doc comment could help explain this (or a need a better name).

File: lib/ASTGen/Sources/ASTGen/Exprs.swift
Comment (ahoppen) [inline lib/ASTGen/Sources/ASTGen/Exprs.swift:55]:
```suggestion
      break
```

File: lib/ASTGen/Sources/ASTGen/Exprs.swift
Comment (ahoppen) [inline lib/ASTGen/Sources/ASTGen/Exprs.swift:60]: I don’t think `current.parent` needs to be non-nil if the entire syntax tree starts with an `ExprSyntax`. I think we should return `true` if there is no more parent.

File: lib/ASTGen/Sources/ASTGen/Exprs.swift
Comment (ahoppen) [inline lib/ASTGen/Sources/ASTGen/Exprs.swift:50]: Is there any value in tracking the known unimplemented kinds? Or is this meant as a To-Do list of nodes to migrate?

File: lib/ASTGen/Sources/ASTGen/LegacyParse.swift
Comment (ahoppen) [inline lib/ASTGen/Sources/ASTGen/LegacyParse.swift:7]: Many whitespaces

File: lib/ASTGen/Sources/ASTGen/LegacyParse.swift
Comment (ahoppen) [inline lib/ASTGen/Sources/ASTGen/LegacyParse.swift:1]: Shouldn’t we have copyright headers in ASTGen?

File: lib/ASTGen/Sources/ASTGen/LegacyParse.swift
Comment (ahoppen) [inline lib/ASTGen/Sources/ASTGen/LegacyParse.swift:17]: I’m not entirely sure what you mean by top expressions here. Maybe an example would help

File: lib/ASTGen/Sources/ASTGen/Types.swift
Comment (ahoppen) [inline lib/ASTGen/Sources/ASTGen/Types.swift:35]: Same comment to `parent` being `nil` here as for expressions.

File: lib/DriverTool/swift_parse_test_main.cpp
Comment (ahoppen) [inline lib/DriverTool/swift_parse_test_main.cpp:63]: What exactly is this supposed to dump? The AST?

File: lib/ASTGen/Sources/ASTGen/LegacyParse.swift
Comment (ahoppen) [inline lib/ASTGen/Sources/ASTGen/LegacyParse.swift:11]: I wonder if this should be `positionAfterSkippingLeadingTrivia` because that corresponds to the start location of the token in the C++ AST.

File: lib/Parse/ParseBridging.cpp
Comment (ahoppen) [inline lib/Parse/ParseBridging.cpp:39]: We somehow also need to propagate `hasError` and `hasCodeCompletionToken`. Is that something you want to do in this PR or in a follow-up?

File: lib/Parse/ParseBridging.cpp
Comment (ahoppen) [inline lib/Parse/ParseBridging.cpp:34]: Do you also want to add the `/*PreviousLoc=*/ annotation here? I think it really helps.

File: lib/Parse/ParseBridging.cpp
Comment (ahoppen) [inline lib/Parse/ParseBridging.cpp:50]: Why do we collect the `decls` here? I don’t think we do anything with them.

File: lib/ASTGen/Sources/ASTGen/Exprs.swift
Comment (rintaro) [reply lib/ASTGen/Sources/ASTGen/Exprs.swift:60]: This is walking up the parents from the first token of `node`. `.parent` must exist if `current` is not `node`

File: lib/ASTGen/Sources/ASTGen/Exprs.swift
Comment (rintaro) [reply lib/ASTGen/Sources/ASTGen/Exprs.swift:50]: Yeah, I consider this is a to-do list. If we have a `ExprSyntax` specific `SyntaxKind` or `SyntaxEnum` it should e great.

File: lib/DriverTool/swift_parse_test_main.cpp
Comment (rintaro) [reply lib/DriverTool/swift_parse_test_main.cpp:63]: Whatever the result. for libParse and ASTGen, the result is AST. For SwiftParser, I'm planning to just dump the syntax tree. It's just convenient to confirm that it actually performed the parsing.

File: lib/Parse/ParseBridging.cpp
Comment (rintaro) [reply lib/Parse/ParseBridging.cpp]: No reason. Removed.

File: lib/ASTGen/Sources/ASTGen/LegacyParse.swift
Comment (rintaro) [reply lib/ASTGen/Sources/ASTGen/LegacyParse.swift]: It doesn't really matter, but you are right. I removed this method. and changed to use the existing `SyntaxProtocol.bridgedSourceLoc(in: ASTGenVisitor)`  method.

File: lib/ASTGen/Sources/ASTGen/LegacyParse.swift
Comment (rintaro) [reply lib/ASTGen/Sources/ASTGen/LegacyParse.swift]: Rephrased.

File: include/swift/Parse/Parser.h
Comment (rintaro) [reply include/swift/Parse/Parser.h:966]: I _think_ we are slowly migrating to "camelCase" for function parameters. I prefer new parameters being lower camel case.

File: include/swift/Parse/Parser.h
Comment (rintaro) [reply include/swift/Parse/Parser.h:163]: I don't think "we" (developers with `SWIFT_BUILD_SWIFT_SYNTAX ` environment) would do much in `#else` branches in Parser.  I'm more worried that developers for non-`SWIFT_BUILD_SWIFT_SYNTAX` environments use `IsForASTGen` mistakingly.

File: lib/ASTGen/Sources/ASTGen/Exprs.swift
Comment (rintaro) [inline lib/ASTGen/Sources/ASTGen/Exprs.swift:38]: Literals are already implemented. I didn't notice they were in `Literals.swift`.

File: lib/ASTGen/Sources/ASTGen/Exprs.swift
Comment (ahoppen) [reply lib/ASTGen/Sources/ASTGen/Exprs.swift:60]: Oh, I see. Could you add a comment that contains exactly what you said in the reply?

File: lib/ASTGen/Sources/ASTGen/Exprs.swift
Comment (ahoppen) [reply lib/ASTGen/Sources/ASTGen/Exprs.swift:50]: 👍🏽 

FWIW, adding an `ExprSyntaxEnum` to swift-syntax would make total sense to me.

File: lib/DriverTool/swift_parse_test_main.cpp
Comment (ahoppen) [reply lib/DriverTool/swift_parse_test_main.cpp:63]: Ah, OK 👍🏽

File: lib/ASTGen/Sources/ASTGen/LegacyParse.swift
Comment (ahoppen) [reply lib/ASTGen/Sources/ASTGen/LegacyParse.swift]: It’s a lot clearer now. Thanks!

File: include/swift/Parse/Parser.h
Comment (ahoppen) [reply include/swift/Parse/Parser.h:966]: Disclaimer: I don’t know what the transition path is here but at least I have always tried to be consistent with the rest of the surrounding code. But I’ll leave it up to you, ultimately.

File: include/swift/Parse/Parser.h
Comment (ahoppen) [reply include/swift/Parse/Parser.h:163]: What do you imagine someone might use `IsForASTGen` for that would be guarded by the `#if`?

Comment (rintaro) [reply include/swift/Parse/Parser.h:163]: I have no idea. But if it's inside `#if SWIFT_BUILD_SWIFT_SYNTAX` they cannot use it by mistake. So I thought it's safer to guard it.
Though I'm totally OK not to guard this, if you think that's better.

**Comment by rintaro** [comment]
@swift-ci Please smoke test

**Comment by rintaro** [comment]
@swift-ci Please smoke test

**Comment by rintaro** [comment]
preset=asan
@swift-ci Please test with preset macOS platform

**Comment by rintaro** [comment]
@swift-ci Please smoke test

**Comment by rintaro** [comment]
preset=asan
@swift-ci Please test with preset macOS platform

**Comment by DougGregor** [comment]
I like this approach of bouncing between the two parsers as needed, so we can opt in to ASTGen production-by-production.

**Comment by rintaro** [comment]
@swift-ci Please smoke test

File: lib/ASTGen/Sources/ASTGen/ASTGen.swift
Comment (rintaro) [reply lib/ASTGen/Sources/ASTGen/ASTGen.swift:558]: Currently `findSyntaxNodeInSourceFile()` prints an error message to stdout, which is not great, but at least it prevents "unwanted surprises". But yeah, I think we should add a diagnostic on the diagnostic engine.
But since this PR is getting bigger, and to unblock some follow-up patches I have, I'd like to do it in the followups.
Also, I'm thinking to change `findSyntaxNodeInSourceFile()` itself. https://github.com/apple/swift/pull/69724#discussion_r1387258659

**Comment by rintaro** [comment]
preset=asan
@swift-ci Please test with preset macOS platform

File: test/ASTGen/types.swift
Comment (rintaro) [inline test/ASTGen/types.swift:5]: It is really unfortunate that this test hits ASAN failure too. We're actively looking into this issue.
Anyway I confirmed these tests passes locally.

Comment (hamishknight) [reply test/ASTGen/types.swift:5]: Should we mark it `UNSUPPORTED: asan` for now so we at least don't lose the test coverage?

File: include/swift/Parse/Parser.h
Comment (ahoppen) [reply include/swift/Parse/Parser.h:163]: It’s fine. If you think it provides value, let’s guard it.

**Review by ahoppen — APPROVED** [review]

File: lib/DriverTool/swift_parse_test_main.cpp
Comment (bnbarham) [inline lib/DriverTool/swift_parse_test_main.cpp:137]: Is this used? Only reference I see is in ParseDecl.cpp

Comment (rintaro) [reply lib/DriverTool/swift_parse_test_main.cpp:137]: Ah, it's a leftover. I'm removing it in a followup

File: include/swift/Parse/Parser.h
Comment (hamishknight) [reply include/swift/Parse/Parser.h:163]: FWIW we have had issues in the past where `#if`'ing out fields leads to crashes if the library and its client ever disagree about the condition (IIRC most recently we had issues with `#ifndef NDEBUG`). Not sure if that can be the case here, but I would tend on the side of not `#if`'ing out fields in general.

**Review by hamishknight — COMMENTED** [review]
Cool!

**Pull request merged**
